### PR TITLE
Add configuration to turn on/off async logging using ActionJob to do it

### LINKED ISSUE #8: Async logging
Allow logging async, add some configuration to turn it on/off and use ActionJob to do it.

### PR DESCRIPTION
**Pull request by matteo95g**
### Summary

Closes #8 - Allow logging async, added configuration to turn it on/off using ActionJob to do it.

### CODE REVIEW

File: spec/dummy/config/initializers/exception_hunter.rb [hunk 1]
```diff
@@ -58,4 +58,11 @@
   #     webhook: SLACK_WEBHOOK_URL_2'
   #   }
   # }
+
+  # == Async Logging
+  #
+  # You can configure if you want to log errors async
+  # Default: true
```
Comment (brunvez) [inline spec/dummy/config/initializers/exception_hunter.rb:65]:
```suggestion
  # Default: false
```

File: spec/error_creator_spec.rb [hunk 2]
```diff
@@ -124,6 +124,28 @@ module ExceptionHunter
               end
             end
           end
+
+          context 'when async logging is set to true' do
+            let!(:original_queue_adapter) { ActiveJob::Base.queue_adapter }
+            let(:async_logging)           { true }
+            let(:tag)                     { nil }
```
Comment (brunvez) [inline spec/error_creator_spec.rb:131]: Please don't align them 🙈

File: spec/error_creator_spec.rb [hunk 3]
```diff
@@ -124,6 +124,28 @@ module ExceptionHunter
...
+            before do
+              ActiveJob::Base.queue_adapter = :test
+
+              allow(ExceptionHunter::Config)
+                .to receive(:async_logging)
+                .and_return(async_logging)
+            end
+
+            after do
+              ActiveJob::Base.queue_adapter = original_queue_adapter
```
Comment (brunvez) [inline spec/error_creator_spec.rb:142]: Do we even need to change it after? I'm not against it, just wondering if there was any reason

File: spec/error_creator_spec.rb (hunk 2 above)
Comment (brunvez) [inline spec/error_creator_spec.rb:131]: Maybe a real tag would be better, even if it does not get used

File: README.md [hunk 4]
```diff
@@ -122,7 +122,7 @@ config.notifiers << {
 }
 ```
 
-6. Add the code below to the environment config file where you are using ExceptionHunter with the correct server url.
+5. Add the code below to the environment config file where you are using ExceptionHunter with the correct server url.
```
Comment (brunvez) [inline README.md:125]: My school teacher would be mad at me, thanks

File: lib/exception_hunter/error_creator.rb [hunk 5]
```diff
@@ -15,6 +16,12 @@ class << self
       def call(tag: nil, **error_attrs)
```
Comment (brunvez) [inline lib/exception_hunter/error_creator.rb:16]: The name `call` doesn't make sense anymore, does it? Perhaps `create_or_schedule_create`? I don't like it either, open to suggestions

File: lib/exception_hunter/error_creator.rb [hunk 6]
```diff
@@ -28,16 +35,26 @@ def call(tag: nil, **error_attrs)
...
         end
-      rescue ActiveRecord::RecordInvalid
-        false
       end
 
       private
 
+      def log_error(tag, error_attrs)
+        if async_logging?
+          ExceptionHunter::AsyncLoggingJob.perform_later(tag, error_attrs)
```
Comment (brunvez) [inline lib/exception_hunter/error_creator.rb:44]: Are we sending the `occurend_at` here? If not that time would be offset when the job is called

File: README.md (hunk 4 above)
Comment (martinjaimem) [inline README.md:125]: doesn't
1.
1.
1.
1.

work in this case?

Comment (brunvez) [reply README.md:125]: Yeah that's a good solution

File: lib/exception_hunter/error_creator.rb (hunk 5 above)
Comment (brunvez) [reply lib/exception_hunter/error_creator.rb:16]: We could also do something like 
```ruby
def call(async: Config.async, tag: nil, ...)
```

Which allows us to inject the config always giving us more flexibility. Your call anyway

File: spec/error_creator_spec.rb (hunk 3 above)
Comment (matteo95g) [reply spec/error_creator_spec.rb:142]: it was not necessary, I removed it

File: README.md [hunk 7]
```diff
@@ -132,7 +132,19 @@ end
 
 This uses ActiveJob to send notification in the background, so [make sure you configure](https://guides.rubyonrails.org/active_job_basics.html#setting-the-backend) it with the adapter you are using, if not notifications will be sent synchronously.
 
+## Async Logging
```
Comment (brunvez) [inline README.md:135]: We could also clarify that it's only async for requests, job errors will be logged sync

File: lib/exception_hunter/error_creator.rb [hunk 8]
```diff
@@ -1,5 +1,7 @@
 module ExceptionHunter
   # Core class in charge of the actual persistence of errors and notifications.
+  # :reek:DataClump
+  # :reek:ControlParameter
```
Comment (brunvez) [inline lib/exception_hunter/error_creator.rb:4]: We generally add these to reek's config

File: lib/exception_hunter/error_creator.rb [hunk 9]
```diff
@@ -12,9 +14,26 @@ class << self
...
+
+      private
+
+      def log_error(tag, error_attrs, async_logging)
+        if async_logging
+          error_attrs = error_attrs.merge(occurred_at: Time.now)
+          ::ExceptionHunter::AsyncLoggingJob.perform_later(tag, error_attrs)
+        else
+          create_error(tag, error_attrs)
+        end
```
Comment (brunvez) [inline lib/exception_hunter/error_creator.rb:33]: Maybe we can move this to `.call`?

File: lib/exception_hunter/error_creator.rb [hunk 10]
```diff
@@ -12,9 +14,26 @@ class << self
...
+      rescue ActiveRecord::RecordInvalid
+        false
+      end
+
+      private
+
+      def log_error(tag, error_attrs, async_logging)
+        if async_logging
+          error_attrs = error_attrs.merge(occurred_at: Time.now)
+          ::ExceptionHunter::AsyncLoggingJob.perform_later(tag, error_attrs)
```
Comment (brunvez) [inline lib/exception_hunter/error_creator.rb:30]: How about
```suggestion
          ::ExceptionHunter::AsyncLoggingJob.perform_later(tag, error_attrs.merge(occurred_at: Time.now))
```

Too crowded? Up to you

File: app/jobs/exception_hunter/async_logging_job.rb [hunk 11]
```diff
@@ -0,0 +1,12 @@
+module ExceptionHunter
+  class AsyncLoggingJob < ApplicationJob
+    queue_as :default
+
+    def perform(tag, error_attrs)
+      ErrorCreator.call(async_logging: false, tag: tag, **error_attrs)
+    rescue Exception # rubocop:disable Lint/RescueException
```
Comment (brunvez) [inline app/jobs/exception_hunter/async_logging_job.rb:7]: Same here, you can disable a cop in a specific file.

**Review by brunvez — COMMENTED** [review]
One last round, don't forget to add yourself to the changelog

**Review by brunvez — APPROVED** [review]

**Pull request merged**